--- ./Makefile
+++ ./Makefile
@@ -25,7 +25,7 @@
 OBJECTS= $(foreach version,$(VERSIONS), \
 					 $(foreach module,$(MODULES), $(module)_$(version).cmo))
 
-METAQUOTS= $(foreach version,$(VERSIONS),ppx_metaquot_$(version))
+METAQUOTS= $(foreach version,$(VERSIONS),ppx_metaquot_$(version)$(EXE))
 .PHONY: all
 all: ppx_tools_versioned.cma ppx_tools_versioned.cmxa $(METAQUOTS)
 
@@ -88,7 +88,7 @@
 
 # Metaquot
 
-ppx_metaquot_%: ppx_tools_versioned.cmxa ast_lifter_%.cmx ppx_metaquot_%.cmx ppx_metaquot_run.cmx
+ppx_metaquot_%$(EXE): ppx_tools_versioned.cmxa ast_lifter_%.cmx ppx_metaquot_%.cmx ppx_metaquot_run.cmx
 	$(OCAMLOPT) -o $@ $(COMPFLAGS) -linkpkg $^
 
 .PHONY: depend
