--- ./myocamlbuild.ml
+++ ./myocamlbuild.ml
@@ -20,6 +20,11 @@
   in
   rule "js_of_ocaml: .byte -> .js" ~dep ~prod f
 
+let lib s =
+  match !Ocamlbuild_plugin.Options.ext_lib with
+  | "" -> s ^ ".a"
+  | x -> s ^ "." ^ x
+
 let () =
   dispatch begin function
   | After_rules ->
@@ -29,8 +34,8 @@
       flag ["link"; "library"; "ocaml"; "native"; "use_mtime"]
         (S ([A "-cclib"; A "-lmtime_stubs"] @ system_support_lib));
       flag ["link"; "ocaml"; "link_mtime"]
-        (A "src-os/libmtime_stubs.a");
+        (A (lib "src-os/libmtime_stubs"));
       dep ["link"; "ocaml"; "use_mtime"]
-        ["src-os/libmtime_stubs.a"];
+        [lib "src-os/libmtime_stubs"];
   | _ -> ()
   end
--- ./src-os/mtime_stubs.c
+++ ./src-os/mtime_stubs.c
@@ -18,6 +18,8 @@
  #if defined(_POSIX_VERSION)
    #define OCAML_MTIME_POSIX
  #endif
+#elif defined(_WIN32)
+#define OCAML_MTIME_WINDOWS
 #endif
 
 /* Darwin */
@@ -63,6 +65,31 @@
                           (uint64_t)(now.tv_nsec - start.tv_nsec));
 }
 
+#elif defined(OCAML_MTIME_WINDOWS)
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#include <stdint.h>
+
+CAMLprim value ocaml_mtime_elapsed_ns (value unit)
+{
+  static LARGE_INTEGER start;
+  static double freq;
+  if (freq == 0.0)
+  {
+    LARGE_INTEGER t_freq;
+    if (!QueryPerformanceFrequency(&t_freq)) return caml_copy_int64(0L);
+    freq = (1000000000.0 / t_freq.QuadPart);
+  }
+  if ( start.QuadPart == 0 )
+  {
+    if (!QueryPerformanceCounter(&start)) return caml_copy_int64(0L);
+  }
+  static LARGE_INTEGER now;
+  if ( !QueryPerformanceCounter(&now)) return caml_copy_int64(0L);
+  uint64_t ret = (now.QuadPart - start.QuadPart) * freq;
+  return caml_copy_int64(ret);
+}
+
 /* Unsupported */
 
 #else
